test/cpp_api_parity/module_impl_check.py

```diff
@@ -0,0 +1,308 @@
+
+    run_cpp_test_fn_and_check_output()
+
+    # Remove temporary folder that stores C++ outputs, but don't block the process
+    # if this fails.
+    try:
+        shutil.rmtree(cpp_tmp_folder)
+    except:  # noqa: E722
+        pass
+
+def test_torch_nn_module_variant(unit_test_class, test_params):
+    test_forward_backward(unit_test_class, test_params)
+
+def compute_module_name(test_params_dict):
+    fullname = test_params_dict.get('fullname', None)
+    if fullname:
+        module_name = fullname.split('_')[0]
+    else:
+        module_name = test_params_dict.get('module_name')
+    return module_name
+
+def process_test_params_for_module(test_params_dict, device, test_instance_class):
+    module_name = compute_module_name(test_params_dict)
+    test_params_dict['constructor'] = test_params_dict.get('constructor', getattr(torch.nn, module_name))
+    test_instance = test_instance_class(**test_params_dict)
+    assert test_instance.get_name().startswith('test_')
+    # Example output: `BCELoss_weights_cuda`
+    module_variant_name = test_instance.get_name()[5:] + (('_' + device) if device != 'cpu' else '')
+
+    if 'constructor_args' in test_params_dict:
+        assert 'cpp_constructor_args' in test_params_dict, (
+            "If `constructor_args` is present in test params dict, to enable C++ API parity test, "
+            "`cpp_constructor_args` must be present in:\n{}"
+            "If you are interested in adding the C++ API parity test, please see:\n"
+            "NOTE [How to check NN module / functional API parity between Python and C++ frontends]. \n"
+            "If not, please add `test_cpp_api_parity=False` to the test params dict and file an issue about this."
+        ).format(pprint.pformat(test_params_dict))
+
+    return TorchNNModuleTestParams(
+        module_name=module_name,
+        module_variant_name=module_variant_name,
+        test_instance=test_instance,
+        cpp_constructor_args=test_params_dict.get('cpp_constructor_args', ''),
+        arg_dict=compute_arg_dict(test_params_dict, test_instance),
+        has_parity=test_params_dict.get('has_parity', True),
+        device=device,
+        cpp_tmp_folder=tempfile.mkdtemp(),
+    )
+
+def add_torch_nn_module_impl_parity_tests(
+        parity_table, unit_test_class, test_params_dicts, test_instance_class, devices):
+    if not hasattr(unit_test_class, 'module_test_params_map'):
+        unit_test_class.module_test_params_map = {}
+    for test_params_dict in test_params_dicts:
+        # Skip all `torch.nn.functional` tests, since they are handled by another test suite.
+        if is_torch_nn_functional_test(test_params_dict):
+            continue
+
+        module_name = compute_module_name(test_params_dict)
+
+        assert hasattr(torch.nn, module_name), \
+            "`torch.nn` doesn't have module `{}`. ".format(module_name) + \
+            "If you are adding a new test, please set `fullname` using format `ModuleName_desc`, " + \
```

Comment:
We should beef this up a little. The question I have while reading this is, "where am I setting these"? Is it in the module test dictionary?